open_spiel/games/mancala.cc

```diff
@@ -0,0 +1,237 @@
+      if(board_[board_.size() - 1 - i] > 0){
+        moves.push_back(board_.size() - 1 - i);
+      }
+    }
+  }
+  return moves;
+}
+
+std::string MancalaState::ActionToString(Player player,
+                                           Action action_id) const {
+  return absl::StrCat(action_id);
+}
+
+void MancalaState::InitBoard() {
+  std::fill(begin(board_), end(board_), 4);
+  board_[0] = 0;
+  board_[board_.size() / 2] = 0;
+}
+
+MancalaState::MancalaState(std::shared_ptr<const Game> game) : State(game) {
+  InitBoard();
+}
+
+std::string MancalaState::ToString() const {
+  std::string str;
+  std::string separator = "-";
+  absl::StrAppend(&str, separator);
+  for (int i = 0; i < kNumPits; ++i) {
+    absl::StrAppend(&str, board_[board_.size() - 1 - i]);
+    absl::StrAppend(&str, separator);
+  }
+  absl::StrAppend(&str, "\n");
+
+  
+  absl::StrAppend(&str, board_[0]);
+  for (int i = 0; i < kNumPits * 2 - 1; ++i) {
+    absl::StrAppend(&str, separator);
+  }
+  absl::StrAppend(&str, board_[board_.size() / 2]);
+  absl::StrAppend(&str, "\n");
+
+
+  absl::StrAppend(&str, separator);
+  for (int i = 0; i < kNumPits; ++i) {
+    absl::StrAppend(&str, board_[i + 1]);
+    absl::StrAppend(&str, separator);
+  }
+  return str;
+}
+
+bool MancalaState::IsTerminal() const {
+  bool player_0_has_moves = false;
+  bool player_1_has_moves = false;
+  for (int i = 0; i < kNumPits; ++i) {
+    if(board_[board_.size() - 1 - i] > 0){
+      player_1_has_moves = true;
+      break;
+    }    
+  }
+  for (int i = 0; i < kNumPits; ++i) {
+    if(board_[i + 1] > 0){
+      player_0_has_moves = true;
+      break;
+    }    
+  }
+  return !player_0_has_moves || !player_1_has_moves;
+}
+
+std::vector<double> MancalaState::Returns() const {
+  int player_0_bean_sum = std::accumulate(board_.begin() + 1, board_.begin() + kTotalPits / 2 + 1, 0);
+  int player_1_bean_sum = std::accumulate(board_.begin() + kTotalPits / 2 + 1, board_.end(), 0) + board_[0];
+  if (player_0_bean_sum > player_1_bean_sum) {
+    return {1.0, -1.0};
+  } else if (player_0_bean_sum < player_1_bean_sum) {
+    return {-1.0, 1.0};
+  } else {
+    return {0.0, 0.0};
+  }
+}
+
+std::string MancalaState::InformationStateString(Player player) const {
+  SPIEL_CHECK_GE(player, 0);
+  SPIEL_CHECK_LT(player, num_players_);
+  return HistoryString();
+}
+
+std::string MancalaState::ObservationString(Player player) const {
+  SPIEL_CHECK_GE(player, 0);
+  SPIEL_CHECK_LT(player, num_players_);
+  return ToString();
+}
+
+void MancalaState::ObservationTensor(Player player,
+                                       absl::Span<float> values) const {
+  SPIEL_CHECK_GE(player, 0);
+  SPIEL_CHECK_LT(player, num_players_);
+
+  SPIEL_CHECK_EQ(values.size(), kTotalPits);
+  auto value_it = values.begin();
+  for (int count : board_) {
+    *value_it++ = count;
+  }
+  SPIEL_CHECK_EQ(value_it, values.end());
+}
+
+void MancalaState::UndoAction(Player player, Action move) {
+  //Yet to implement Undo
```

Comment:
Either remove the comment, or implement the function. Many of the games do not have Undo, so you can just remove the function if it's not correct.

If you leave it, please add one of the simulation tests that tests the undo: https://github.com/deepmind/open_spiel/blob/63e8a99b70a4ea2eb934740d8fb1d2add2dd258b/open_spiel/tests/basic_tests.h#L57